Framework/Algorithms/inc/MantidAlgorithms/SpectrumAlgorithm.h

```diff
@@ -0,0 +1,65 @@
+#ifndef MANTID_ALGORITHMS_SPECTRUMALGORITHM_H_
+#define MANTID_ALGORITHMS_SPECTRUMALGORITHM_H_
+
+#include <tuple>
+
+#include "MantidKernel/SpectrumIndexSet.h"
+#include "MantidAPI/Algorithm.h"
+#include "MantidAPI/MatrixWorkspace_fwd.h"
+#include "MantidAlgorithms/DllConfig.h"
+#include "MantidAlgorithms/Getters.h"
+
+namespace Mantid {
+
+namespace Kernel {
+class SpectrumIndexSet;
+}
+
+namespace Algorithms {
+
+class MANTID_ALGORITHMS_DLL SpectrumAlgorithm : public API::Algorithm {
+private:
+  // 3 little helpers for generating a sequence 1,2,3,.... Used for extracting
+  // arguments from tuple.
+  template <int...> struct seq {};
+  template <int N, int... S> struct gens : gens<N - 1, N - 1, S...> {};
+  template <int... S> struct gens<0, S...> { typedef seq<S...> type; };
+
+  template <class WS, class T, int... S, class OP>
+  void for_each(WS &workspace, T getters, seq<S...>, const OP &operation) {
+    //auto count = static_cast<int64_t>(workspace.getNumberHistograms());
+
+    auto indexSet = getSpectrumIndexSet(workspace);
```

Comment:
I think I agree with your comment about the fact this it is a little hidden in here. From the callers point of you it just looks like it will go over the whole workspace so a change of name would be good I think.

Do we think there are cases where it would be useful for the caller of `for_each` to be able to also specify a range explicitly? Admittedly this is more likely to just introduce more problems with consistent behaviour.